### ISSUE TITLE
System.IO.FileLoadException when multiple target is set to netcoreapp2.0 and net461

### ISSUE THREAD

**Issue by viper3400** [GregFinzer/Compare-Net-Objects#102]
I attach a Visual Studio 2017 solution for better understanding. I've set up a console app which targets two frameworks: netcoreapp2.0 and net461.

[Jaxx.CompareNetCore.zip](https://github.com/GregFinzer/Compare-Net-Objects/files/1996080/Jaxx.CompareNetCore.zip)

This console app references a netstandard2.0 library. This library references Compare-NET-Objects NuGet pacakge, Version 4.53.0.

Publish the app from the console app directory via

```
dotnet publish -f netcoreapp2.0
```
and
```
dotnet publish -f net461
```

When running the netcoreapp2.0 all is fine, when running the net461 exe I've got the following exception:

```
System.IO.FileLoadException
  HResult=0x80131040
  Message=Die Datei oder Assembly "KellermanSoftware.Compare-NET-Objects-Core, Version=4.54.0.0, Culture=neutral, PublicKeyToken=dccbd7ce7d6a58c0" oder eine Abhängigkeit davon wurde nicht gefunden. Die gefundene Manifestdefinition der Assembly stimmt nicht mit dem Assemblyverweis überein. (Ausnahme von HRESULT: 0x80131040)
  Source=Jaxx.CompareNetCore.Shared
  StackTrace:
   at Jaxx.CompareNetCore.Shared.Comparer.CompareEqual() in D:\WORK\_Playground\Jaxx.CompareNetCore\Jaxx.CompareNetCore.Shared\Comparer.cs:line 12
   at Jaxx.CompareNetCore.ConsoleApp.Program.Main(String[] args) in D:\WORK\_Playground\Jaxx.CompareNetCore\Jaxx.CompareNetCore.Console\Program.cs:line 13
```

Obviously framework searches for Compare-NET-Objects-Core.dll - which is not deployed o the project. (Instead Compare-NET-Objects.dll is deployed to into the net461 build)

A bit confusing too is file version of Compare-NET-Objects.dll, it's 4.53.0 while version of Compare-NET-Objects-Core.dll is 4.54.0.
 
![grafik](https://user-images.githubusercontent.com/11006334/39934831-68df996a-5547-11e8-92c7-4ab27a3f4a62.png)

Found #94 , but this seems not to be the same.

Thanks for help.

**Comment by GregFinzer** [comment]
@viper3400 I am not sure what to do about that.  Any suggestions?

**Comment by viper3400** [comment]
Hi @GregFinzer 

Had a deeper look into it and I think it's a problem of the NuGet package. Have you tried to reproduce this problem with my test solution?

I've cloned this repo and build it on my own. I've checked the "Generate NuGet package on build" in "Package" tab in properties of Compare-NET-Objects-Core project and used this package:

[Compare-NET-Objects-Core-Test.4.54.0.zip](https://github.com/GregFinzer/Compare-Net-Objects/files/1997374/Compare-NET-Objects-Core-Test.4.54.0.zip)

This works for me. When switching back to your package the error reoccurs.

I've made a diff of both deployments and found some differences in the exe.config file and some differences in the files (resp. in the files naming) which are deployed.

```
-- Only in my
Jaxx.CompareNetCore/Jaxx.CompareNetCore.Console/bin/Debug/net461/
  Jaxx.CompareNetCore.Console.exe.config 
Jaxx.CompareNetCoreWORKS/Jaxx.CompareNetCore.Console/bin/Debug/net461/
  Jaxx.CompareNetCore.Console.exe.config

18,23d17
<         <assemblyIdentity name="System.Runtime.Serialization.Json" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
<         <bindingRedirect oldVersion="0.0.0.0-4.0.1.0" newVersion="4.0.1.0" />
<       </dependentAssembly>
<     </assemblyBinding>
<     <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
<       <dependentAssembly>

137a132,137
>         <assemblyIdentity name="System.Console" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
>         <bindingRedirect oldVersion="0.0.0.0-4.0.2.0" newVersion="4.0.2.0" />
>       </dependentAssembly>
>     </assemblyBinding>
>     <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
>       <dependentAssembly>

240a241,246
>         <bindingRedirect oldVersion="0.0.0.0-4.0.1.0" newVersion="4.0.1.0" />
>       </dependentAssembly>
>     </assemblyBinding>
>     <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
>       <dependentAssembly>
>         <assemblyIdentity name="System.Runtime.Serialization.Json" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />

-- Only in my
Only in Jaxx.Jaxx.CompareNetCoreWORKS/Jaxx.CompareNetCore.Console/bin/Debug/net461/: KellermanSoftware.Compare-NET-Objects-Core.dll
Only in Jaxx.Jaxx.CompareNetCoreWORKS/Jaxx.CompareNetCore.Console/bin/Debug/net461/: System.Reflection.TypeExtensions.dll

-- Only in yours
Only in Jaxx.CompareNetCore/Jaxx.CompareNetCore.Console/bin/Debug/net461/: KellermanSoftware.Compare-NET-Objects.dll
```

So there seems to be some dangle with the dll when building the .NET Core app on multiple frameworks. I don't know how you've generated the NuGet package?

**Comment by viper3400** [comment]
Hi @GregFinzer 

I compared your package and the package I've created from your Compare-NET-Objects-Core project:
1st) In my package are two additional dependencies
2nd) In my package there is just the netstandard1.3 dll

![grafik](https://user-images.githubusercontent.com/11006334/39966164-769a94ae-56a7-11e8-885a-562de8efb95f.png)

As you know by now my .NET Core 2.0 project targets two frameworks, the .NET Core 2.0 framework itself and .NET 4.6.1

Excerpt from my .csproj file:

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFrameworks>netcoreapp2.0;net461</TargetFrameworks>
  </PropertyGroup> 
</Project>
```

When using my package the build process creates the right dll version and resolves dependencies by itself for the .NET 4.6.1 project.

If NuGet package provides an own folder (resp. version) for the target framework (as your package does), build process seems to deploy the files in there.

I assume that you build the files in netstandard1.3 folder in your NuGet package from the "Compare-NET-Objects-Core" project and all other binaries from "Compare-NET-Objects" project?

My .NET Core project expects the binary and its dependencies from the "Compare-NET-Objects-Core" build (gracefully for debugging it has another name "KellermanSoftware.Compare-NET-Objects-Core.dll") but gets the binary from your "Compare-NET-Objects-Core"  project, which targets the "old" .NET Frameworks natively "KellermanSoftware.Compare-NET-Objects.dll").

I think, if you'd like to provide a package which supports .NET Standard you need to build all binaries from "Compare-NET-Objects-Core" project by adding the desired frameworks to .csproj file.

https://www.bartwolff.com/Blog/2017/03/20/targeting-multiple-net-platforms-in-a-single-nuget-package-w

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFrameworks>netstandard1.3;net46</TargetFrameworks>
  </PropertyGroup> 
</Project>
```

I'm not the Microsoft guy and that is just how I understood my current problem with the package. So this multiple target thing seems to get a bit messy ;)  I also haven't tested the backwards compatibility of the dll created from the .NET Core project in a native .NET 4.6 application. Maybe this confusion leads to separate NuGet packages for .NET Core targets at all ...

**Comment by viper3400** [comment]
Hi @GregFinzer 

I tried, but that won't change the behaviour, because the net46 folder is deployed to my build folder, but my build depends on "KellermanSoftware.Compare-NET-Objects-Core.dll".

So I could add the "KellermanSoftware.Compare-NET-Objects-Core.dll" and the depended "System.Reflection.TypeExtension.dll" to the net46 folder, but I think this would be just a hack.

The content of the netstandard2.0 folder is deployed when I publish the netcore2.0 app.

![grafik](https://user-images.githubusercontent.com/11006334/39967652-af4aa2f4-56bf-11e8-812d-6b261ae681b5.png)

dotnet publish -f net461

![grafik](https://user-images.githubusercontent.com/11006334/39967653-b32e9a6a-56bf-11e8-88c5-2cf4f7a05f37.png)

dotnet publish -f netcoreapp2.0

![grafik](https://user-images.githubusercontent.com/11006334/39967654-b60c1d66-56bf-11e8-8fbf-4b60fe5762f8.png)

**Comment by GregFinzer** [comment]
@viper3400 Thanks for your help.  It looks like this is going to take several hours to fix.  It will be several weeks until I can get to this.

**Comment by jasase** [comment]
We have also the same problem. We think one problem could be that the dll for .Net Standard has another name than the other dlls. Every other NuGet package that target multiple .Net Framework version use the same DLL name for every version.

**Comment by GregFinzer** [comment]
@jasase Good information.  Thanks.

**Comment by twurm** [comment]
@jasase the fix outlined in #108 and the associated pull request fixed the issue for me.  If you need a short term fix you can pull that and build it . Also thanks to @ltrzesniewski for figuring it out.

**Comment by GregFinzer** [comment]
@viper3400 @jasase @ltrzesniewski 
This should be fixed in the latest build.
